js/src/evaluation/_runner.ts

```diff
@@ -0,0 +1,932 @@
+      project = this._experiment;
+    }
+    return project;
+  }
+
+  _printExperimentStart(): void {
+    // @TODO log with experiment URL
+    console.log(`Starting evaluation of experiment: ${this.experimentName}`);
+  }
+}
+
+/**
+ * Manage the execution of experiments.
+ *
+ * Supports lazily running predictions and evaluations in parallel to facilitate
+ * result streaming and early debugging.
+ */
+class _ExperimentManager extends _ExperimentManagerMixin {
+  _data?: DataT;
+
+  _examples?: AsyncGenerator<Example>;
+
+  _runs?: AsyncGenerator<Run>;
+
+  _evaluationResults?: AsyncGenerator<EvaluationResults>;
+
+  _summaryResults?: AsyncGenerator<
+    (runsArray: Run[]) => AsyncGenerator<EvaluationResults, any, unknown>,
+    any,
+    unknown
+  >;
+
+  _examplesArray?: Example[];
+
+  _runsArray?: Run[];
+
+  constructor(args: _ExperimentManagerArgs) {
+    super({
+      experiment: args.experiment,
+      metadata: args.metadata,
+      client: args.client,
+    });
+    if (args._examplesArray && args._examplesArray.length) {
+      this._examplesArray = args._examplesArray;
+    }
+    this._data = args.data;
+
+    if (args._runsArray && args._runsArray.length) {
+      this._runsArray = args._runsArray;
+    }
+    this._runs = args.runs;
+
+    this._evaluationResults = args.evaluationResults;
+    this._summaryResults = args.summaryResults;
+  }
+
+  get examples(): AsyncGenerator<Example> {
+    if (this._examples === undefined) {
+      if (!this._data) {
+        throw new Error("Data not provided in this experiment.");
+      }
+      return _resolveData(this._data, { client: this.client });
+    }
+    return this._examples;
+  }
+
+  get datasetId(): Promise<string> {
+    if (!this._experiment || !this._experiment.reference_dataset_id) {
+      const examplesIterator = this.examples[Symbol.asyncIterator]();
+      return examplesIterator.next().then((result) => {
+        if (result.done) {
+          throw new Error("No examples found in the dataset.");
+        }
+        return result.value.dataset_id;
+      });
+    }
+    return Promise.resolve(this._experiment.reference_dataset_id);
+  }
+
+  get evaluationResults(): AsyncGenerator<EvaluationResults> {
+    if (this._evaluationResults === undefined) {
+      return async function* (this: _ExperimentManager) {
+        for (const _ of this._examplesArray ?? []) {
+          yield { results: [] };
+        }
+      }.call(this);
+    } else {
+      return this._evaluationResults;
+    }
+  }
+
+  get runs(): AsyncGenerator<Run> {
+    if (this._runsArray && this._runsArray.length > 0) {
+      throw new Error("Runs already provided as an array.");
+    }
+    if (this._runs === undefined) {
+      throw new Error(
+        "Runs not provided in this experiment. Please predict first."
+      );
+    } else {
+      return this._runs;
+    }
+  }
+
+  async start(): Promise<_ExperimentManager> {
+    if (!this._examplesArray) {
+      this._examplesArray = [];
```

Comment:
Maybe makes sense to just add this within the `examples` getter  as the cache so you don't have 2 ways of accessing it

Reply:
`examples` getter has a return type of `AsyncGenerator<Example>` so if we want to add it there we either need to convert it to a generator, or add a union return type and add checks for the type everywhere it's used

Reply:
Removed async generator examples getter, replaced with an async `getExamples` method